File: scripts/parse-sheet.js
import { readFileSync, writeFileSync } from "fs";

const raw = JSON.parse(
  readFileSync("/vercel/share/v0-project/scripts/sheet.json", "utf-8")
);

const questions = raw.data.questions;

// Build topic -> subTopic -> questions hierarchy preserving insertion order
const topicMap = new Map();

for (const q of questions) {
  const topicName = q.topic || "Uncategorized";
  const subTopicName = q.subTopic || "General";

  if (!topicMap.has(topicName)) {
    topicMap.set(topicName, new Map());
  }
  const subMap = topicMap.get(topicName);
  if (!subMap.has(subTopicName)) {
    subMap.set(subTopicName, []);
  }

  const difficulty = (q.questionId?.difficulty || "Medium").toLowerCase();
  let normalizedDifficulty = "medium";
  if (difficulty === "easy" || difficulty === "basic") normalizedDifficulty = "easy";
  else if (difficulty === "hard") normalizedDifficulty = "hard";

  subMap.get(subTopicName).push({
    title: q.title,
    difficulty: normalizedDifficulty,
    link: q.questionId?.problemUrl || "",
    resource: q.resource || "",
    platform: q.questionId?.platform || "",
  });
}

const TOPIC_COLORS = [
  "#f97316", "#ea580c", "#fb923c", "#c2410c", "#f59e0b",
  "#d97706", "#ef4444", "#f43f5e", "#e11d48", "#ec4899",
  "#10b981", "#14b8a6", "#3b82f6", "#6366f1", "#8b5cf6",
];

const result = [];
let colorIdx = 0;

for (const [topicName, subMap] of topicMap) {
  const subTopics = [];
  for (const [subTopicName, qs] of subMap) {
    subTopics.push({
      title: subTopicName,
      questions: qs,
    });
  }
  result.push({
    title: topicName,
    color: TOPIC_COLORS[colorIdx % TOPIC_COLORS.length],
    subTopics,
  });
  colorIdx++;
}

console.log("Parsed " + result.length + " topics");
let totalQ = 0;
for (const t of result) {
  let tq = 0;
  for (const st of t.subTopics) tq += st.questions.length;
  console.log("  " + t.title + ": " + t.subTopics.length + " sub-topics, " + tq + " questions");
  totalQ += tq;
}
console.log("Total questions: " + totalQ);

const output = `// Auto-generated from sheet.json - Do not edit manually

export interface SheetQuestion {
  title: string;
  difficulty: "easy" | "medium" | "hard";
  link: string;
  resource: string;
  platform: string;
}

export interface SheetSubTopic {
  title: string;
  questions: SheetQuestion[];
}

export interface SheetTopic {
  title: string;
  color: string;
  subTopics: SheetSubTopic[];
}

export const SHEET_DATA: SheetTopic[] = ${JSON.stringify(result, null, 2)};
`;

writeFileSync("/vercel/share/v0-project/lib/sheet-data.ts", output);
console.log("Written to lib/sheet-data.ts");

File: scripts/parse-sheet.py
import json
import os
import glob

# Find the sheet json file
sheet_path = None
for pattern in [
    "/vercel/share/v0-project/scripts/sheet.json",
    "/vercel/share/v0-project/scripts/sheet-*.json",
]:
    matches = glob.glob(pattern)
    if matches:
        sheet_path = matches[0]
        break

if not sheet_path:
    # Try to find it anywhere
    for root, dirs, files in os.walk("/vercel/share/v0-project"):
        for f in files:
            if f.startswith("sheet") and f.endswith(".json"):
                sheet_path = os.path.join(root, f)
                break

print(f"Looking for sheet at: {sheet_path}")
print(f"File exists: {os.path.exists(sheet_path) if sheet_path else False}")

# List what's in scripts dir
scripts_dir = "/vercel/share/v0-project/scripts"
if os.path.exists(scripts_dir):
    print(f"Scripts dir contents: {os.listdir(scripts_dir)}")

# Also check current working dir
print(f"CWD: {os.getcwd()}")
print(f"CWD contents: {os.listdir('.')}")

if not sheet_path or not os.path.exists(sheet_path):
    print("ERROR: sheet.json not found!")
    exit(1)

with open(sheet_path, "r") as f:
    raw = json.load(f)

questions = raw["data"]["questions"]
print(f"Total raw questions: {len(questions)}")

# Build topic -> subTopic -> questions hierarchy
from collections import OrderedDict

topic_map = OrderedDict()

TOPIC_COLORS = [
    "#f97316", "#ea580c", "#fb923c", "#c2410c", "#f59e0b",
    "#d97706", "#ef4444", "#f43f5e", "#e11d48", "#ec4899",
    "#10b981", "#14b8a6", "#3b82f6", "#6366f1", "#8b5cf6",
    "#f97316", "#ea580c", "#fb923c", "#c2410c", "#f59e0b",
    "#d97706", "#ef4444", "#f43f5e", "#e11d48", "#ec4899",
    "#10b981", "#14b8a6",
]

for q in questions:
    topic_name = q.get("topic", "Uncategorized")
    sub_topic_name = q.get("subTopic", "General")
    
    if topic_name not in topic_map:
        topic_map[topic_name] = OrderedDict()
    sub_map = topic_map[topic_name]
    if sub_topic_name not in sub_map:
        sub_map[sub_topic_name] = []
    
    qid = q.get("questionId") or {}
    difficulty_raw = (qid.get("difficulty") or "Medium").lower()
    if difficulty_raw in ("easy", "basic"):
        difficulty = "easy"
    elif difficulty_raw == "hard":
        difficulty = "hard"
    else:
        difficulty = "medium"
    
    sub_map[sub_topic_name].append({
        "title": q.get("title", "Untitled"),
        "difficulty": difficulty,
        "link": qid.get("problemUrl", ""),
        "resource": q.get("resource", ""),
        "platform": qid.get("platform", ""),
    })

result = []
color_idx = 0
for topic_name, sub_map in topic_map.items():
    sub_topics = []
    for sub_name, qs in sub_map.items():
        sub_topics.append({
            "title": sub_name,
            "questions": qs,
        })
    result.append({
        "title": topic_name,
        "color": TOPIC_COLORS[color_idx % len(TOPIC_COLORS)],
        "subTopics": sub_topics,
    })
    color_idx += 1

total_q = 0
for t in result:
    tq = sum(len(st["questions"]) for st in t["subTopics"])
    print(f"  {t['title']}: {len(t['subTopics'])} sub-topics, {tq} questions")
    total_q += tq

print(f"\nTotal: {len(result)} topics, {total_q} questions")

# Generate TypeScript
lines = []
lines.append("// Auto-generated from sheet.json - Do not edit manually\n")
lines.append("export interface SheetQuestion {")
lines.append("  title: string;")
lines.append("  difficulty: \"easy\" | \"medium\" | \"hard\";")
lines.append("  link: string;")
lines.append("  resource: string;")
lines.append("  platform: string;")
lines.append("}\n")
lines.append("export interface SheetSubTopic {")
lines.append("  title: string;")
lines.append("  questions: SheetQuestion[];")
lines.append("}\n")
lines.append("export interface SheetTopic {")
lines.append("  title: string;")
lines.append("  color: string;")
lines.append("  subTopics: SheetSubTopic[];")
lines.append("}\n")
lines.append(f"export const SHEET_DATA: SheetTopic[] = {json.dumps(result, indent=2)};")

output = "\n".join(lines) + "\n"

out_path = "/vercel/share/v0-project/lib/sheet-data.ts"
with open(out_path, "w") as f:
    f.write(output)

print(f"\nWritten to {out_path} ({len(output)} bytes)")
